### ISSUE TITLE
Aliasing on canvas.Line

### ISSUE THREAD

**Issue by charlesdaniels** [fyne-io/fyne#1246]
### Describe the bug:

Noticeable "jaggies" appear on a `canvas.Line` object drawn at an angle.

### To Reproduce:

Run the attached example code.

### Screenshots:

![screenshot_2020-08-11_at_13_26_26](https://user-images.githubusercontent.com/11738895/89928739-47a98e00-dbd6-11ea-9e34-8f6b6e72e7e8.png)

It may be hard to tell if your browser is scaling the image, but there are noticeable jaggies on the line.

### Example code:

```go
package main

import (
	"fyne.io/fyne"
	"fyne.io/fyne/app"
	"fyne.io/fyne/canvas"
	"fyne.io/fyne/theme"
	"fyne.io/fyne/widget"
	"image/color"
)

type fooRenderer struct {
	line fyne.CanvasObject
}

func (f *fooRenderer) MinSize() fyne.Size {
	return fyne.NewSize(100, 100)
}

func (f *fooRenderer) Layout(fyne.Size) {}

func (f *fooRenderer) ApplyTheme() {}

func (f *fooRenderer) BackgroundColor() color.Color {
	return theme.BackgroundColor()
}

func (f *fooRenderer) Refresh() {
	f.line.(*canvas.Line).Position1 = fyne.Position{0, 0}
	f.line.(*canvas.Line).Position2 = fyne.Position{100, 20}
	f.line.(*canvas.Line).StrokeColor = theme.TextColor()
	f.line.(*canvas.Line).StrokeWidth = 5
	canvas.Refresh(f.line)
}

func (f *fooRenderer) Destroy() {
}

func (f *fooRenderer) Objects() []fyne.CanvasObject {
	return []fyne.CanvasObject{f.line}
}

type FooWidget struct {
	widget.BaseWidget
}

func (w *FooWidget) Tapped(ev *fyne.PointEvent) {
}

func (w *FooWidget) TappedSecondary(ev *fyne.PointEvent) {
}

func (w *FooWidget) CreateRenderer() fyne.WidgetRenderer {
	f := &fooRenderer{
		canvas.NewLine(theme.TextColor()),
	}
	f.Refresh()

	return f
}

func NewFooWidget() *FooWidget {
	f := &FooWidget{}
	f.ExtendBaseWidget(f)
	return f
}

func main() {
	a := app.New()
	w := a.NewWindow("Hello")

	w.SetContent(widget.NewVBox(NewFooWidget()))

	w.ShowAndRun()
}
```

### Device (please complete the following information):
 - OS: Linux
 - Version: Ubuntu 20.04, kernel 5.4.0-42-generic, x86_64
 - Go version: 1.13.8 linux/amd64
 - Fyne version: f6a908307cd87cc36364aebf59085f27fe784b82
    - I can also reproduce against 1.3.0 and 1.3.3 (have not tested other versions)

**Comment by andydotxyz** [comment]
I have zoomed in on the image and drawn a single pixel in red. I exploded it 5* to show the detail.
It looks anti-aliased to me. Perhaps you can be more detailed in your expectation?
![89928739-47a98e00-dbd6-11ea-9e34-8f6b6e72e7e8](https://user-images.githubusercontent.com/294436/89932743-06c96d80-dc06-11ea-907c-d255a2338389.png)

**Comment by andydotxyz** [comment]
P.S. smaller code to reproduce the issue:

```go
package main

import (
	"fyne.io/fyne"
	"fyne.io/fyne/app"
	"fyne.io/fyne/canvas"
	"fyne.io/fyne/theme"
)

func main() {
	a := app.New()
	w := a.NewWindow("Hello")

	line := canvas.NewLine(theme.TextColor())
	line.Position1 = fyne.Position{0, 0}
	line.Position2 = fyne.Position{100, 20}
	line.StrokeColor = theme.TextColor()
	line.StrokeWidth = 5

	w.SetContent(fyne.NewContainer(line))
	w.Resize(fyne.NewSize(120, 50))

	w.ShowAndRun()
}
```

**Comment by charlesdaniels** [comment]
Agreed, there's definitely some AA going on there. However, I think it looks more jaggy than other programs. For example, consider this SVG:

```html
<!DOCTYPE html>
<html>
<body>

<svg height="210" width="500">
  <line x1="10" y1="10" x2="110" y2="30" style="stroke:rgb(0,0,0);stroke-width:5" />
  Sorry, your browser does not support inline SVG.
</svg>

</body>
</html>
```
(You can "run" it [here](https://www.w3schools.com/graphics/tryit.asp?filename=trysvg_line))

![screenshot_2020-08-11_at_14_38_20](https://user-images.githubusercontent.com/11738895/89935606-6ca2fe80-dbe0-11ea-8af7-8b1ed52528e3.png)

Looking at it side-by-side, I think the one in the SVG looks smoother.

**Comment by andydotxyz** [comment]
Subtle, but I agree.
Perhaps there are options to rasterx’s dasher? 
https://github.com/fyne-io/fyne/blob/6c9ed1b5aa618a55ef3051aca3f6806bdc74dbf7/internal/painter/draw.go#L53

**Comment by andydotxyz** [comment]
I looked at the alternative FreeType based antialiaser in srwiley's code - but I don't think it makes much difference.
<img width="865" alt="Screenshot 2020-08-12 at 10 31 15" src="https://user-images.githubusercontent.com/294436/89999690-0a510900-dc87-11ea-84f3-4bcad4b44140.png">

 To move this further we would need to find another rasterising platform I think.

**Comment by andydotxyz** [comment]
Actually what we need to do is move this to OpenGL renderers, this bug report maybe moot once that is done

**Comment by charlesdaniels** [comment]
Checked with #2217 applied, and I don't see any change. Attached is the result from running my original sample code.

![screenshot_2021-05-09_at_22_09_47](https://user-images.githubusercontent.com/11738895/117597487-ac692900-b113-11eb-9290-019d117176fc.png)

**Comment by andydotxyz** [comment]
It looks to me a little better - but I agree it's marginal if at all.
At this point I am not sure what to do. A custom anti-alias algorithm is not really an option once it's moved to the GPU is it?

**Comment by charlesdaniels** [comment]
I would assume that it is in fact possible for GPUs to perform AA.

I don't really consider this issue super high priority, but I think once we have a more sophisticated OpenGL back-end, it would be a nice-to-have.

**Comment by andydotxyz** [comment]
> I would assume that it is in fact possible for GPUs to perform AA.

If no AA was applied I would agree, but as discussed above it is successfully being anti-aliased, the difference is in the specific alias algorithm that can be applied?

**Comment by charlesdaniels** [comment]
I would assume, without being an OpenGL expert, that you can tune the AA to your heart's content, given that the game dev people  have quite an array of different effects available.

Whether this is something practical for us to actually implement, that's another matter entirely.
